### PR TITLE
AzureMonitor: strongly-typed AzureCredentials and correct resolution of auth type and cloud

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

This PR even further abstracts out AzureTokenProvider from Azure Monitor datasource by introducing strongly-typed credentials structs `AzureManagedIdentityCredentials` and `AzureClientSecretCredentials` of common interface `AzureCredentials`.

The logic which parses auth params (`JsonData`) was moved out of `AzureTokenProvider` to the level of the datasource. As result, `AzureTokenProvider` doesn't enforce datasource schema and parsing rules and each datasource can store credentials in its own way. Corner cases like when `authType` not specified but `clientId` is specified are not part of the token provider logic anymore, as they specific to backward compatibility of Azure Monitor datasource.

`AzureCredentials` and `AzureTokenProvider` now use globally known cloud names like `AzureChinaCloud` (consistent with server config), rather than Azure Monitor specific cloud names like `chinaazuremonitor`.

There are few regressions were fixed:
1. In case of managed identity, cloud name should be used from server configuration, not from JsonData
2. If datasource has no configuration (empty JsonData), it's still valid datasource and cloud name should be properly implied
3. Mismatches between global cloud names from server configuration and Azure Monitor cloud names from JsonData

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

http client factory was moved from `credentials.go` to `httpclient.go`, and `credentials.go` only contain code for parsing of AzureCredentials.

Parsing logic in `credentials.go` covered with unit tests.

I still have a few unit tests to fix.

Packages `azcredentials` and `aztokenrpovider` will be eventually moved into a shared repo and the package structure will be like this:

```go
import (
    ...

	"github.com/grafana/grafana-azure-sdk-go/pkg/azcredentials"
	"github.com/grafana/grafana-azure-sdk-go/pkg/aztokenprovider"
)

...

credentials := &azcredentials.AzureManagedIdentityCredentials{}
provider := aztokenprovider.NewAzureAccessTokenProvider(cfg, credentials)
```

### CODE REVIEW

File: pkg/tsdb/azuremonitor/azcredentials/credentials.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package azcredentials
+
+const (
+	AzureAuthManagedIdentity = "msi"
+	AzureAuthClientSecret    = "clientsecret"
+)
+
+type AzureCredentials interface {
+	AzureAuthType() string
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/azcredentials/credentials.go:9]: maybe repeating `Azure` is not necessary? --> `AuthType() string`

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go [hunk 2]
```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
...
+		scopes:      scopes,
 	}
 }
 
 func (provider *tokenProviderImpl) GetAccessToken(ctx context.Context) (string, error) {
 	var credential TokenCredential
 
-	if provider.isManagedIdentityCredential() {
+	// TODO: Move to provider initialization and reuse for each GetAccessToken call
+	switch c := provider.credentials.(type) {
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:40]: IMO it's a bit more readable if you use the `AuthType` method for the switch based on a string and do the conversion later on

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go [hunk 3]
```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
...
-	// * If authType and other fields aren't set then it means the datasource never been configured
-	//   and managed identity is the default authentication choice as long as managed identities are enabled
-	return authType == "msi" || (authType == "" && clientId == "" && provider.cfg.Azure.ManagedIdentityEnabled)
-}
-
-func (provider *tokenProviderImpl) getManagedIdentityCredential() TokenCredential {
-	clientId := provider.cfg.Azure.ManagedIdentityClientId
-
-	return &managedIdentityCredential{clientId: clientId}
+func (provider *tokenProviderImpl) getManagedIdentityCredential(credentials *azcredentials.AzureManagedIdentityCredentials) TokenCredential {
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:63]: not related to this PR but I find a bit confusing to have `TokenCredential` and `azcore.TokenCredential` both being different interfaces. As far as I can tell the difference is that this `TokenCredential` manages a cache, so I'd suggest to use `CachedTokenCredential` to make it clear.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go [hunk 4]
```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
...
+	var clientId string
+	if credentials.ClientId != "" {
+		clientId = credentials.ClientId
+	} else {
+		clientId = provider.cfg.Azure.ManagedIdentityClientId
+	}
+	return &managedIdentityCredential{
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:70]: also unrelated to this PR but why having an `Init` method as part of the interface if we could do that in this method? (reducing the risk of calling `GetAccessToken` without the struct being initialized.)

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go [hunk 5]
```diff
@@ -93,9 +98,9 @@ func (provider *tokenProviderImpl) resolveAuthorityHost(cloudName string) string
 		return azidentity.AzureGovernment
 	case setting.AzureGermany:
 		return azidentity.AzureGermany
+	default:
+		return ""
 	}
-	// Fallback to direct URL
-	return provider.authParams.Url
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:98]: why this change?

File: pkg/tsdb/azuremonitor/azuremonitor_test.go [hunk 6]
```diff
@@ -29,22 +28,25 @@ func TestNewInstanceSettings(t *testing.T) {
...
-			factory := NewInstanceSettings()
-			instance, err := factory(tt.settings)
+			factory := NewInstanceSettings(cfg)
+			_, err := factory(tt.settings)
 			tt.Err(t, err)
-			if !cmp.Equal(instance, tt.expectedModel, cmpopts.IgnoreFields(datasourceInfo{}, "Services", "HTTPCliOpts")) {
-				t.Errorf("Unexpected instance: %v", cmp.Diff(instance, tt.expectedModel))
-			}
+			// TODO: Fix comparison
+			//if !cmp.Equal(instance, tt.expectedModel, cmpopts.IgnoreFields(datasourceInfo{}, "Services", "HTTPCliOpts")) {
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/azuremonitor_test.go:47]: You can add more fields to the ignored fields if that's what was failing for you

**Review by andresmgot — COMMENTED** [review]
Mostly looking good for me. Just a few question/minor suggestions. Thanks for the tests!

File: pkg/tsdb/azuremonitor/azuremonitor_test.go (hunk 6 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/azuremonitor_test.go]: Tried many ways and could not make it work :( I reverted back JSONData to old type for now. It doesn't make functional difference.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go (hunk 5 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:98]: While we didn't actually use it, I initially implemented that  `authParams.Url` was used to provide custom IdP authorization endpoint, e.g. for unknown cloud or for testing. But it was implemented the way that value from `authParams.Url` was used only if cloud didn't match.

In the new implementation `AzureClientSecretCredentials` struct has fields `AzureCloud` and `Authority`:

```go
type AzureClientSecretCredentials struct {
	AzureCloud   string
	Authority    string
	...
}
```

If `Authority` isn't set (as is the normal case) then default authority used for a given cloud `AzureCloud`.
If `Authority` is set, then it overrides default authority from the cloud.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go (hunk 4 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go:66]: `TokenCredential` struct serves two purposes:
- as a cache key accessor
- as an abstraction layer over the GetAccessToken implementation (Azure SDK's `azidentity`).

In the old implementation of token provider `TokenCredential` was created on every call of `GetAccessToken()`  and in majority of cases was thrown away unused after checking the key because `ConcurrentTokenCache` always use the instance of `TokenCredential` which it has already in cache.

`Init` method has `azidentity` initialization which is expensive, for example `azidentity.NewManagedIdentityCredential` does probing of endpoints to determine managed identities implementation. So I didn't want to have expensive initialization only to throw out the struct. Only initialization of the struct fields is cheap.

`ConcurrentTokenCache` will call `Init` only in case if it actually intends to use it.

In the new implementation I have moved the `TokenCredential` creation to constructor to make implementation of `AzureTokenProvider` even more lightweight. In this case `Init` is useful to make sure that `azidentity` being initialized during the `GetAccessToken` call, not during construction of `AzureTokenProvider`.   Init can fail, and I want it to fail during GetAccessToken, to properly propagate the error as result of `GetAccessToken`.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go (hunk 3 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go]: Agree, the naming is confusing.

Initially I wrote the `ConcurrentTokenCache` as generic cache, to related to Azure SDK.
Another thing is initialization of `azcore.TokenCredential` is expensive, so I wanted a lightweight wrapper with lazy initialization via `Init`.

I will rename it.

File: pkg/tsdb/azuremonitor/azcredentials/credentials.go (hunk 1 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/azcredentials/credentials.go:9]: Go automatically matches structs to interfaces and there are other interfaces which implement `AuthType() string` :)
So I just made it unique enough to not clash with other interfaces.

My goal is to implement discriminated union type: `AzureCredentials` which is either `AzureManagedIdentityCredentials` or `AzureClientSecretCredentials`, basically [this](https://github.com/grafana/grafana/blob/b255f3db3f84d897f33deedb063e6ea5fd86eada/public/app/plugins/datasource/grafana-azure-monitor-datasource/types/index.ts#L74),

I saw on internet that people usually add a dummy method to interface like this:
```go
type MyUnionType interface {
     ...

     isMyUnionType() // dummy method to bind interface
}
```
Just for a purpose of binding the interface to implementation structs.

Example here:
https://medium.com/@haya14busa/sum-union-variant-type-in-go-and-static-check-tool-of-switch-case-handling-3bfc61618b1e

I decided to utilize the `AzureAuthType`, it has some practical use and also used to "glue" the interface to implementation instead of a dummy method.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go (hunk 2 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go]: I just used the popular pattern for sum types / discriminated unions:
[here](https://making.pusher.com/alternatives-to-sum-types-in-go/) and [here](https://medium.com/@haya14busa/sum-union-variant-type-in-go-and-static-check-tool-of-switch-case-handling-3bfc61618b1e)

Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go]: in this case `c` is automatically cased to the proper type in each `case` branch

**Comment by kostrse** [comment]
New changes:

I added two new fields into the `datasourceInfo` struct: `Cloud` and `Credentials`:

```go
type datasourceInfo struct {
	Cloud       string
	Credentials azcredentials.AzureCredentials
	...
}
```

The `Cloud` field properly computed based on type of authentication and server config and can be reliably used when it's necessary to know the cloud of the datasource (instead of `Settings.CloudName`).

I deleted credentials-related fields from the `Settings` just to make sure they are not used anywhere. They are not reliable, the `Credentails` has the correct credentials if needed.

**Comment by kostrse** [comment]
`scopes []string` were moved from `AzureTokenProvider` constructor to the `GetAccessToken` method:

```go
type AzureTokenProvider interface {
	GetAccessToken(ctx context.Context, scopes []string) (string, error)
}

func NewAzureAccessTokenProvider(*setting.Cfg, azcredentials.AzureCredentials) (AzureTokenProvider, error)
```

Azure Management API uses the same scope for all REST API calls, so for Azure Monitor it is sufficient to provide the scope once into constructor and then use the HTTP client for different HTTP requests with the same scope, so the original parameterless `GetAccessToken` worked well. But it's rather exception than the rule.

Normally, different REST API methods have different scopes, e.g.:
`GET /user` can require scope `User.Read`, `POST /user` requires scope `User.Write`, `GET /users` requires scope `User.GetAll` etc.

So, proper semantic is to initialize HTTP client and token provider with `credentials`, and then reuse it for different HTTP requests by providing proper `scopes` on the per-request basis.

Because `AzureTokenProvider` is supposed to be shared with other datasources and other Microsoft APIs, I enforced this semantic. For Azure Monitor it doesn't make much practical difference.
The only small optimization Azure Monitor datasource can have is it doesn't need to create a new instance of AzureTokenProvider for each executor, it can share it, since they all share the same credentials. Executors have different scopes but the scopes are passed now to `GetAccessToken`.

File: pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go (hunk 3 above)
Comment (kostrse) [reply pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go]: Renamed to `TokenRetriever`. Wanted to call `TokenAccessor` but `AccessToken` vs `TokenAccessor` sounds weird :)
I decided to not call it `CachedTokenCredential` because it actually doesn't cache the token, it's accessor / getter. Maybe could call `getter` but doesn't metter, it's internal implementation which is not exposed out of the package,

**Review by andresmgot — APPROVED** [review]
> I deleted credentials-related fields from the Settings just to make sure they are not used anywhere. They are not reliable, the Credentails has the correct credentials if needed.

Accepting already but I just have a comment following up with this. Thanks for the insights @kostrse !

BTW, I have asked internally about the code scanning alert you are getting here to solve it. [Edit] Done.

**Comment by kostrse** [comment]
Let's merge it, it's getting large.

We can address remaining stuff in another PR:
1. SameAs as a calculated field which takes into account auth type (Managed Identity cannot have different credentials)
2. JSONData as simplejson.Json* to avoid unnecessary conversion
3. SecureJsonData probably not needed, AppInsights credentials can be a struct similar to Credentials
4. AzureTokenProvider can be shared among executors

**Comment by andresmgot** [comment]
> Let's merge it, it's getting large.
> 
> We can address remaining stuff in another PR:
> 
> 1. SameAs as a calculated field which takes into account auth type (Managed Identity cannot have different credentials)
> 2. JSONData as simplejson.Json* to avoid unnecessary conversion
> 3. SecureJsonData probably not needed, AppInsights credentials can be a struct similar to Credentials
> 4. AzureTokenProvider can be shared among executors

Sounds good. Thanks!
